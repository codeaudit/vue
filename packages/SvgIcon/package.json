{
  "name": "@versionone/ui-svg-icon",
  "version": "0.0.0",
  "ui/name": "SvgIcon",
  "ui/status": "experimental",
  "description": "",
  "main": "dist/index.js",
  "files": [
    "dist"
  ],
  "ui/webpack/entry": "src/index.js",
  "ui/webpack/output": "dist/index.js",
  "scripts": {
    "prepublish": "{ cd ../../; yarn run prepublish/single @versionone/ui-svg-icon; };"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/versionone/vue.git"
  },
  "config": {
    "access": "public"
  },
  "keywords": [
   "icon",
    "svg"
  ],
  "author": "VersionOne",
  "maintainers": [
    "Andrew Smith <andrew@andrew.codes>"
  ],
  "license": "MIT",
  "peerDependencies": {
    "react": "15.1.0"
  },
  "dependencies": {
    "@versionone/ui-utilities": "0.0.0"
  }
}
