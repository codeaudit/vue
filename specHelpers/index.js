import React from 'react';
import ThemeProvider, {getTheme} from './../src/Theme';

export const applyTheme = (ComponentUnderTest, theme = getTheme()) => <ThemeProvider theme={theme}>{ComponentUnderTest}</ThemeProvider>;
