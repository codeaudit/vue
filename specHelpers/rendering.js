import toJson from 'enzyme-to-json';
import {mount, shallow} from 'enzyme';
import React from 'react';
import theme from './TestTheme';

export const getRender = (Component) => (props = {}, context = {theme}) => shallow(<Component {...props} />, {context});
export const getMount = (Component) => (props = {}, context = {theme}) => mount(<Component {...props} />, {context});
export const snapshot = (renderedComponent) => toJson(renderedComponent);
