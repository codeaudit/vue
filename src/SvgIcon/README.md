The SvgIcon component takes an SVG `path` element as its child; converting it to a React component which displays the path. This component also enables the icon to be styled and respond to mouse events. SvgIcon components are always squares, so setting the width will also set the same value as the height.

## VersionOne Icons
VersionOne UI also includes icons used by VersionOne. Each of these icons is its own component. More information on these icons can be found on the [Icons' page](#/page/icons).
