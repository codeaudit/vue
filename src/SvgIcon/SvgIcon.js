import React, {PropTypes,} from 'react';
import TrackingHover from './../utilities/TrackingHover';
import Radium from './../utilities/Radium';

const getStyles = (props) => {
    const {
        color,
        hovered,
        hoverColor,
        transition,
        width,
    } = props;

    return {
        root: {
            fill: hovered ? hoverColor : color,
            height: `${width}px`,
            margin: 0,
            padding: 0,
            transition,
            width: `${width}px`,
        },
    };
};

const defaultTransition = 'fill 0.25s linear 0ms';
const defaultProps = {
    color: '#000',
    transition: defaultTransition,
    width: 24,
};
const SvgIcon = (props) => {
    const propsWithDefaults = {
        ...defaultProps,
        ...props,
    };
    const {children,} = propsWithDefaults;
    const styles = getStyles(propsWithDefaults);
    const eventHandlerProps = Object.keys(propsWithDefaults)
        .filter(key => key.startsWith('on'))
        .reduce((output, key) => ({
            ...output,
            [key]: propsWithDefaults[key],
        }), {});

    return (
        <svg
            {...eventHandlerProps}
            style={styles.root}
            viewBox="0 0 24 24"
            x="0px"
            y="0px"
        >
            {children}
        </svg>
    );
};
SvgIcon.propTypes = {
    /**
     * SVG element's body to be rendered.
     */
    children: PropTypes.node.isRequired,
    /**
     * Fill color of the SVG.
     */
    color: PropTypes.string,
    /**
     *
     */
    hovered: PropTypes.bool,
    /**
     * Fill color to apply when hovering over the SVG.
     */
    hoverColor: PropTypes.string,
    /**
     * Transition to apply to SVG; typically used for fill color on hover.
     */
    transition: PropTypes.string,
    /**
     * Width of the SVG; also applies this as the height.
     */
    width: PropTypes.number,
    /**
     * onClick event handler
     */
    onClick: PropTypes.func,
    /**
     * onMouseEnter event handler.
     */
    onMouseEnter: PropTypes.func,
    /**
     * onMouseLeave event handler.
     */
    onMouseLeave: PropTypes.func,
};

export default Radium(TrackingHover(SvgIcon));
