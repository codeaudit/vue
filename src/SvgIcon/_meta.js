import examples from './examples';

export default {
    name: 'SvgIcon',
    keywords: ['svg', 'icon'],
    'see also': [],
    readme: require('./README.md'),
    status: 'experimental',
    componentsSources: [
        {
            name: 'SvgIcon',
            code: require('!raw!./SvgIcon')
        }
    ],
    examples
};
