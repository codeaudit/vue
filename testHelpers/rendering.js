import toJson from 'enzyme-to-json';
import React, {PropTypes} from 'react';
import {mount, render, shallow} from 'enzyme';
import theme from './../packages/themes/src/V1Theme';

const ctx = {
    theme,
};

export const getShallow = (Component) => (props = {}, context = {}) => shallow(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getRender = (Component) => (props = {}, context = {}) => render(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
});

export const getMount = (Component) => (props = {}, options = {}, context = {}) => mount(<Component {...props} />, {
    childContextTypes: {
        theme: PropTypes.object,
    },
    context: {
        ...ctx,
        ...context,
    },
    ...options,
});

const replaceUIKey = (componentName) => {
    let index = 0;
    return (json) => {
        if (!json.props || !json.props.uiKey) {
            return json;
        }
        const uiKey = `${componentName}-${index}`;
        index += 1;
        return Object.assign({}, json, {
            uiKey,
            uiPath: [uiKey],
        });
    };
};
const replaceUiKeyRecursively = (json) => {
    if (!json.type) {
        return json;
    }
    const replaceUIKeyInParent = replaceUIKey(json.type);
    const output = replaceUIKeyInParent(json);
    if (!output.children) {
        return json;
    }
    output.children = json.children.map((child, index) => replaceUiKeyRecursively(child));
    return output;
};

export const snapshot = (renderedComponent) => {
    return toJson(renderedComponent);
};

export const reset = (component) => () => {
    if (component) {
        component.unmount();
    }
    document.body.innerHTML = '';
};
