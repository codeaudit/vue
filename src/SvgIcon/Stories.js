import React from 'react';
import {storiesOf} from '@kadira/storybook';
import SvgIcon from './';
import AddIcon from './../Icons/Add';

storiesOf('SvgIcon')
    .addWithInfo('custom add svg icon',
        ``,
        () => (
            <SvgIcon>
                <path
                    d="M12,4c-4.4,0-8,3.6-8,8s3.6,8,8,8s8-3.6,8-8S16.4,4,12,4z M17,13h-4.1v3.9h-2V13H7v-2h3.9V6.9h2V11H17V13z"
                />
            </SvgIcon>
        )
    )
    .addWithInfo('AddIcon',
        ``,
        () => (
            <AddIcon />
        )
    )
    .addWithInfo('AddIcon with custom props',
        ``,
        () => (
            <AddIcon
                color="green"
                hoverColor="blue"
                width={50}
            />
        )
    );
