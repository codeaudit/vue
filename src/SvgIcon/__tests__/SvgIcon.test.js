import React from 'react';
import SvgIcon from './../SvgIcon';
import {getShallow, reset, snapshot} from './../../../specHelpers/rendering';

const shallowRenderSvgIcon = getShallow(SvgIcon);

let component;
afterEach(reset(component));

test('will render an SVG element with its children and the correct viewbox value of `0 0 24 24`', () => {
    component = shallowRenderSvgIcon({
        children: getSvgPath(),
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('SvgIcons can have a color', () => {
    component = shallowRenderSvgIcon({
        children: getSvgPath(),
        color: 'blue',
        transition: 'fill 0.5s linear 0ms',
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('SvgIcons can be set to be in a hovered state', () => {
    component = shallowRenderSvgIcon({
        children: getSvgPath(),
        hoverColor: 'blue',
        hovered: true,
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('SvgIcons can be hovered', () => {
    component = shallowRenderSvgIcon({
        children: getSvgPath(),
        hoverColor: 'blue',
    });
    simulateMouseEnter(component);
    expect(snapshot(component)).toMatchSnapshot();
    simulateMouseLeave(component);
    expect(snapshot(component)).toMatchSnapshot();
});

test('SvgIcons can have a width and are always rendered a squares', () => {
    const component = shallowRenderSvgIcon({
        children: getSvgPath(),
        width: 100,
    });
    expect(snapshot(component)).toMatchSnapshot();
});

test('allows standard events to be handled by event handlers', () => {
    const onClick = jest.fn();
    const component = shallowRenderSvgIcon({
        children: getSvgPath(),
        onClick,
    });
    const evt = {test: true};
    simulateClick(component, evt);
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(onClick).toHaveBeenCalledWith(evt);
});

// --
function getSvgPath() {
    return (
        <path
            d="M12,4c-4.4,0-8,3.6-8,8s3.6,8,8,8s8-3.6,8-8S16.4,4,12,4z M17,13h-4.1v3.9h-2V13H7v-2h3.9V6.9h2V11H17V13z"
        />
    );
}
function simulateMouseEnter(wrapper) {
    wrapper.simulate('mouseenter');
}
function simulateMouseLeave(wrapper) {
    wrapper.simulate('mouseleave');
}
function simulateClick(wrapper, evt = {}) {
    wrapper.simulate('click', evt);
}
