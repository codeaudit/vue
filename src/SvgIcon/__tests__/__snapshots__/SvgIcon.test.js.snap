exports[`test SvgIcons can be hovered 1`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "#000",
      "height": "24px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.25s linear 0ms",
      "width": "24px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px" />
`;

exports[`test SvgIcons can be hovered 2`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "#000",
      "height": "24px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.25s linear 0ms",
      "width": "24px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px" />
`;

exports[`test SvgIcons can be set to be in a hovered state 1`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "blue",
      "height": "24px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.25s linear 0ms",
      "width": "24px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px" />
`;

exports[`test SvgIcons can have a color 1`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "blue",
      "height": "24px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.5s linear 0ms",
      "width": "24px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px" />
`;

exports[`test SvgIcons can have a width and are always rendered a squares 1`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "#000",
      "height": "100px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.25s linear 0ms",
      "width": "100px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px" />
`;

exports[`test will render an SVG element with its children and the correct viewbox value of \`0 0 24 24\` 1`] = `
<svg
  data-radium={true}
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "fill": "#000",
      "height": "24px",
      "margin": 0,
      "padding": "0px",
      "transition": "fill 0.25s linear 0ms",
      "width": "24px",
    }
  }
  viewBox="0 0 24 24"
  x="0px"
  y="0px">
  <path
    d="M12,4c-4.4,0-8,3.6-8,8s3.6,8,8,8s8-3.6,8-8S16.4,4,12,4z M17,13h-4.1v3.9h-2V13H7v-2h3.9V6.9h2V11H17V13z" />
</svg>
`;
